### PR TITLE
[JENKINS-37792] Per stage post build steps

### PR DESCRIPTION
**Pull request by rsandell**
@reviewbybees

### CODE REVIEW

**Comment by abayer** [comment]
Yaaaay! =) Reviewing now!

**Comment by rsandell** [comment]
I was planning to clean up all those `c.delegate = script ....` parts as well, but not sure if that should go into this PR or another one. This PR adds yet another one of those and I'm not a fan of redundant code :)

File: src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/ast/ModelASTBuildConditionsContainer.groovy [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+        JSONArray a = new JSONArray()
+        conditions.each { c ->
+            a.add(c.toJSON())
+        }
+
+        return new JSONObject()
+                .accumulate("conditions", a)
+    }
+
+    protected void _validate(ModelValidator validator) {
```
Comment (abayer) [inline src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/ast/ModelASTBuildConditionsContainer.groovy]: I was going to ask why `_validate` and not just `super.validate(...)` calls, but I'm guessing this is to make sure it actually works right for each extension of the class?

Comment (abayer) [inline src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/ast/ModelASTBuildConditionsContainer.groovy]: Works for me.

File: src/test/java/org/jenkinsci/plugins/pipeline/modeldefinition/PostStageTest.java [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+
+    @Test
+    public void failingNotifications() throws Exception {
+        expect(Result.FAILURE, "failingNotifications")
+                .logContains("hello", "Post stage", "farewell", "Error in stage post:", "I AM HERE STILL")
+                .logNotContains("world").go();
+    }
+
+
+    //Helper methods and builders, could possibly be moved to something more central for reuse
```
Comment (abayer) [inline src/test/java/org/jenkinsci/plugins/pipeline/modeldefinition/PostStageTest.java]: Ok, I change my mind, :bug: until this awesome stuff is moved to `AbstractModelDefTest`. =)

**Comment by abayer** [comment]
Now to persuade you to go rewrite all the other tests to use this goodness later... =) :bee:

**Comment by rsandell** [comment]
@reviewbybees done

**Comment by marcusphi** [comment]
It seems this doesn't work all the way. 
The tests in this PR are too simplistic. If you would add a test of case as described in https://issues.jenkins.io/browse/JENKINS-57801 I think you would find that it fails.

**Pull request merged**